src/buf_writer.c

```diff
@@ -5,32 +5,128 @@
 #include "jemalloc/internal/buf_writer.h"
 #include "jemalloc/internal/malloc_io.h"
 
+static void *
+buf_writer_allocate_internal_buf(tsdn_t *tsdn, size_t buf_len) {
+	if (buf_len > SC_LARGE_MAXCLASS) {
+		return NULL;
+	}
+	return iallocztm(tsdn, buf_len, sz_size2index(buf_len), false, NULL,
+	    true, arena_get(tsdn, 0, false), true);
+}
+
+static void
+buf_writer_free_internal_buf(tsdn_t *tsdn, void *buf) {
+	if (buf != NULL) {
```

Comment:
not: Assert is internal.

Reply:
You mean to pass in the `buf_writer` pointer here? My intention was that `buf_writer_allocate_internal_buf()` and `buf_writer_free_internal_buf()` only deal with the buffer array itself, and their respective callers, `buf_writer_init()` and `buf_writer_terminate()`, operate at the `buf_writer` level.

Reply:
ah got auto corrected. I meant: add an assert(buf_is_internal).

Reply:
Discussed offline. Will not add the assert, since `buf_writer_terminate()` tested it before the call, and this function is more meant to be a utility just dealing with the buffer array.